### PR TITLE
Add ws_string_set_from_raw()

### PR DESCRIPTION
**Pull request by xVin**
targets #194

### CODE REVIEW

**Comment by xVin** [comment]
Kindly review. 

(I dunno why this PR has no compiler check yet o.o)

**Comment by matthiasbeyer** [comment]
@xVin look here: http://ci.neikos.me/github.com/waysome/waysome

File: src/objects/string.c [hunk 1]
```diff
@@ -294,6 +294,42 @@ ws_string_raw(
     return output;   
 }
 
+void
+ws_string_set_from_raw(
+    struct ws_string* self,
+    char* raw
+){  
+    if (self) {
```
Comment (neithernut) [inline src/objects/string.c]: I think it would be cleaner to do an early return here.

File: src/objects/string.c [hunk 2]
```diff
@@ -294,6 +294,44 @@ ws_string_raw(
...
+
+    conv_raw = u_strFromUTF8(conv_raw, len + 1, &dest_len, raw, len, &err);
+        
+    if (U_FAILURE(err)) {
+        return;       
+    }
+
+    ws_object_lock_write(&self->obj);
+
+    self->str = realloc(self->str, (len + 1) * sizeof(*self->str));
```
Comment (neithernut) [inline src/objects/string.c]: I do not get why you do the realloc here. Shouldn't `self->str` simply be `free()`d and set to `conv_raw`?

Comment (xVin) [inline src/objects/string.c]: True.

File: src/objects/string.c [hunk 3]
```diff
@@ -294,6 +294,44 @@ ws_string_raw(
...
+    UChar* conv_raw = calloc(len + 1, sizeof(*conv_raw));
+        
+    if (!conv_raw) {
+        return;
+     }
+    
+    int32_t dest_len;
+    UErrorCode err;
+
+    conv_raw = u_strFromUTF8(conv_raw, len + 1, &dest_len, raw, len, &err);
```
Comment (matthiasbeyer) [inline src/objects/string.c:316]: I don't get why the function is called `strFromIUTF8` if it does (or is intended to do) `UTF8FromStr` - are you sure this is correct here?

Comment (xVin) [inline src/objects/string.c:316]: This funtion generates a UChar string from an UTF-8 string (= char*), so it should be what we want here.

File: src/objects/string.c [hunk 4]
```diff
@@ -294,6 +294,44 @@ ws_string_raw(
...
+        
+    if (U_FAILURE(err)) {
+        return;       
+    }
+
+    ws_object_lock_write(&self->obj);
+
+    free(self->str);
+        
+    if (self->str) {
```
Comment (matthiasbeyer) [inline src/objects/string.c]: This is misleading: You do a `free()` beforehand, you should do a `self->str = NULL` there, too. This if would then result in beeing `false` always.

I guess the `if` must be removed.

Comment (xVin) [inline src/objects/string.c]: Oops, forgot that in the previous fix. Will be fixed in a jiffy :o

File: src/objects/string.c (hunk 3 above)
Comment (neithernut) [inline src/objects/string.c:316]: Yep.

**Comment by neithernut** [comment]
ACK.

**Pull request merged**